datalad/cmd.py

```diff
@@ -692,3 +693,151 @@ def run(self, cmd, env=None, *args, **kwargs):
+        # kwargs = dict(bufsize=1, universal_newlines=True) if PY3 else {}
+        self._stderr_out, self._stderr_out_fname = tempfile.mkstemp()
+        self._process = subprocess.Popen(
+            self.cmd,
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=self._stderr_out,
+            env=GitRunner.get_git_environ_adjusted(),
+            cwd=self.path,
+            bufsize=1,
+            universal_newlines=True  # **kwargs
+        )
+
+    def _check_process(self, restart=False):
+        """Check if the process was terminated and restart if restart
+
+        Returns
+        -------
+        bool
+          True if process was alive.
+        str
+          stderr if any recorded if was terminated
+        """
+        process = self._process
+        ret = True
+        ret_stderr = None
+        if process and process.poll():
+            lgr.warning("Process %s was terminated with returncode %s" % (process, process.returncode))
+            ret_stderr = self.close(return_stderr=True)
+            ret = False
+        if self._process is None and restart:
+            lgr.warning("Restarting the process due to previous failure")
+            self._initialize()
+        return ret, ret_stderr
+
+    def __call__(self, cmds):
+        """
+
+        Parameters
+        ----------
+        cmds : str or tuple or list of (str or tuple)
+
+        Returns
+        -------
+        str or list
+          Output received from annex.  list in case if cmds was a list
+        """
+        # TODO: add checks -- may be process died off and needs to be reinitiated
+        if not self._process:
+            self._initialize()
+
+        input_multiple = isinstance(cmds, list)
+        if not input_multiple:
+            cmds = [cmds]
+
+        output = []
+
+        for entry in cmds:
+            if not isinstance(entry, string_types):
+                entry = ' '.join(entry)
+            entry = entry + '\n'
+            lgr.log(5, "Sending %r to batched command %s" % (entry, self))
+            # apparently communicate is just a one time show
+            # stdout, stderr = self._process.communicate(entry)
+            # according to the internet wisdom there is no easy way with subprocess
+            self._check_process(restart=True)
+            process = self._process  # _check_process might have restarted it
+            process.stdin.write(assure_bytes(entry) if PY2 else entry)
+            process.stdin.flush()
+            lgr.log(5, "Done sending.")
+            still_alive, stderr = self._check_process(restart=False)
+            # TODO: we might want to handle still_alive, e.g. to allow for
+            #       a number of restarts/resends, but it should be per command
+            #       since for some we cannot just resend the same query. But if
+            #       it is just a "get"er - we could resend it few times
+            # We are expecting a single line output
```

Comment:
I think this was true for annex batch commands, but it's not true for batch commands in general.  I'll push a clean-up commit that clarifies that the default `output_proc` assumes a single line of output.

Reply:
Cool!